{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172c0ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import os\n",
    "import json\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f9cd240",
   "metadata": {},
   "source": [
    "# LB1 in KDDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e75e6a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "density_dat = pd.read_csv('tables_out/all_dat_tog.tsv', sep='\\t').replace('EarlySomite','Early somite')\n",
    "\n",
    "density_dat['score_lb1'] = density_dat[['score0_lb1','score1_lb1']].mean(axis=1)\n",
    "density_dat['score_k9'] = density_dat[['score0_k9','score1_k9']].mean(axis=1)\n",
    "\n",
    "f, axB = plt.subplots(figsize=(6,8))\n",
    "\n",
    "b = sns.violinplot(x='score_lb1', y='cell_type', data=density_dat, kind='violin',\n",
    "           hue='category_k9',linewidth=0.5, inner=None, ax=axB,\n",
    "                   hue_order=['T1-KDD','T2-KDD','nonKDD'],\n",
    "          order=['ESCs','Mid-hindgut','Liver','Anterior primitive streak',\n",
    "                'Paraxial mesoderm','Early somite','Artery progenitors',d\n",
    "                'Endothelial progenitors','Cardiac myocytes','Epicardium',\n",
    "                'Definitive ectoderm','Border ectoderm','Midbrain progenitors'],\n",
    "                  palette = mypal)\n",
    "plt.legend(frameon=False, title='KDD category', bbox_to_anchor=(1, 1.15),\n",
    "          ncol=1)\n",
    "# plt.xticks(rotation=45, ha='right')\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "axB.set(ylabel='cell type', xlabel='LB1 binding [log2(read counts)]')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f504508",
   "metadata": {},
   "source": [
    "# Proportion expressed genes in LADs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7dbff6",
   "metadata": {},
   "source": [
    "A and B are in .ai file already (track view and violins)\n",
    "\n",
    "C: gene expression\n",
    "\n",
    "D: gene density / ATAC-seq / B compartment enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "209dbdf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_path = 'hg38/kallisto_indices/homo_sapiens/'\n",
    "lads_path = 'LAD_BEDs/'\n",
    "lad_dat = 'LADs/dat'\n",
    "path_to_gene_results = 'LADs/dat/genes'\n",
    "path_to_atac = 'LADs/dat/ATACseq'\n",
    "\n",
    "## gene expression\n",
    "\n",
    "rnaseq_dat = pd.read_table('RNAseq_quantile_normalized_dat.tsv',\n",
    "                          header=0, names=['name','tpm_CardiacMyocytes','tpm_H9ESC',\n",
    "                                           'tpm_EarlySomite','tpm_ParaxMesoderm',\n",
    "                                           'tpm_MidHindgut','tpm_EndoProgenitor',\n",
    "                                           'tpm_DefEctoderm'])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# get locations of genes\n",
    "\n",
    "# get LAD annotations per gene\n",
    "\n",
    "# Kallisto analysis used Ensembl gene annotations v96, should do that here as well for continuity\n",
    "\n",
    "ensembl = pd.read_csv(os.path.join(gtf_path,'Homo_sapiens.GRCh38.96.gtf'),\n",
    "                     sep='\\t', skiprows=5, header=None, \n",
    "                      names=['chrom','source','desc','start','end','score','strand','score2','desc_further'],\n",
    "                     low_memory=False)\n",
    "ensembl['chr'] = 'chr' + ensembl['chrom'].astype(str)\n",
    "ensembl['name'] = ensembl['desc_further'].str.split('gene_name ').str[1].str.split(';').str[0].str.strip('\"\"')\n",
    "\n",
    "# filter for genes only\n",
    "\n",
    "ensembl_genes = ensembl.query('desc == \"gene\"').copy()\n",
    "n_ens_genes = pd.DataFrame(ensembl_genes['name'].value_counts())\n",
    "n_ens_genes['gene'] = n_ens_genes.index\n",
    "ensembl_genes_single = n_ens_genes.query('name == 1')['gene'].tolist()\n",
    "ensembl_genes_single_df = ensembl_genes.query('(name in @ensembl_genes_single)').copy()\n",
    "\n",
    "# make BED file\n",
    "\n",
    "ens_single_bed = BedTool.from_dataframe(ensembl_genes_single_df[['chr','start','end','name']].copy()).sort()\n",
    "\n",
    "\n",
    "\n",
    "def get_lad_state(row):\n",
    "    if row['T1-LAD']:\n",
    "        return('T1-LAD')\n",
    "    elif row['T2-LAD']:\n",
    "        return('T2-LAD')\n",
    "    else:\n",
    "        return('nonLAD')\n",
    "\n",
    "dfs = []\n",
    "\n",
    "for val in rnaseq_dat.columns[1:]:\n",
    "    ct = val.split('_')[1]\n",
    "    t1lads = BedTool(os.path.join(lads_path, f'{ct}_T1-LADs.bed'))\n",
    "    genes_in_t1lads = ens_single_bed.intersect(t1lads, f=1.0).to_dataframe()['name'].drop_duplicates().tolist()\n",
    "    t2lads = BedTool(os.path.join(lads_path, f'{ct}_T2-LADs.bed'))\n",
    "    genes_in_t2lads = ens_single_bed.intersect(t2lads, f=1.0).to_dataframe()['name'].drop_duplicates().tolist()\n",
    "    df = rnaseq_dat[['name', f'tpm_{ct}']].copy()\n",
    "    df.columns = ['name','expression']\n",
    "    df['T1-LAD'] = df['name'].isin(genes_in_t1lads)\n",
    "    df['T2-LAD'] = df['name'].isin(genes_in_t2lads)\n",
    "    df['LAD_state'] = df.apply(lambda row: get_lad_state(row), axis=1)\n",
    "    df['celltype'] = ct\n",
    "    dfs.append(df)\n",
    "    \n",
    "tog = pd.concat(dfs, sort=False)\n",
    "\n",
    "# look at proportion of expressed genes\n",
    "\n",
    "prop_exp = []\n",
    "cats = []\n",
    "cts = []\n",
    "\n",
    "for ct in tog['celltype'].unique():\n",
    "    for cat in tog['LAD_state'].unique():\n",
    "        prop_exp.append(len(tog.query('celltype == @ct and LAD_state == @cat and expression > 5')) / len(tog.query('celltype == @ct and LAD_state == @cat')))\n",
    "        cats.append(cat)\n",
    "        cts.append(ct)\n",
    "        \n",
    "\n",
    "df = pd.DataFrame({\n",
    "    'proportion_expressed':prop_exp,\n",
    "    'LAD_cat':cats,\n",
    "    'celltype':cts\n",
    "})\n",
    "\n",
    "mypal = {'T1-LAD':'indigo','T2-LAD':'mediumpurple','nonLAD':'lightgrey'}\n",
    "\n",
    "f, ax = plt.subplots(figsize=(4,3))\n",
    "sns.barplot(x='LAD_cat', y='proportion_expressed', data=df,\n",
    "           order=['T1-LAD','T2-LAD','nonLAD'], ci='sd',\n",
    "           palette = mypal)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5e1fac4",
   "metadata": {},
   "source": [
    "# Gene, ATAC-peaks, B compartment enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "559b472e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## feature enrichments\n",
    "\n",
    "def get_enrichment_df(enr_results_path, enr_feature):\n",
    "    \"\"\"\n",
    "    Collects enrichment results (pre-computed) into a dataframe for plotting\n",
    "    \"\"\"\n",
    "    zscores = []\n",
    "    pvalues = []\n",
    "    lad_cts = []\n",
    "    enr_elements = []\n",
    "    ladtypes = []\n",
    "\n",
    "    for ct in cts:\n",
    "\n",
    "        enr_dat_t1lad = pd.read_csv(f'{enr_results_path}/{enr_feature}_{ct}_T1LADs_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "        enr_dat_t2lad = pd.read_csv(f'{enr_results_path}/{enr_feature}_{ct}_T2LADs_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "\n",
    "        with open(f'{enr_results_path}/{enr_feature}_{ct}_T1LADs_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t1lads = json.load(f)\n",
    "\n",
    "        if actual_ints_t1lads == 0:\n",
    "            actual_ints_t1lads = 0.00001\n",
    "\n",
    "        with open(f'{enr_results_path}/{enr_feature}_{ct}_T2LADs_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t2lads = json.load(f)\n",
    "\n",
    "        if actual_ints_t2lads == 0:\n",
    "            actual_ints_t2lads = 0.00001\n",
    "\n",
    "        # this loop was originally developed for TEs, but now \n",
    "        # works with all the features we assessed for enrichment\n",
    "        for te_class in enr_dat_t1lad['TE_class'].unique():\n",
    "\n",
    "            # calculate z-score for T1-LAD\n",
    "\n",
    "            t1lad_sd = np.std(enr_dat_t1lad.query('TE_class == @te_class')['n_intersections'])\n",
    "            t1lad_ints = actual_ints_t1lads[te_class]\n",
    "            t1lad_obs_z = (t1lad_ints - \\\n",
    "                         np.mean(enr_dat_t1lad.query('TE_class == @te_class')['n_intersections'])) / t1lad_sd\n",
    "\n",
    "            # calculate empirical p-value for T1-LAD\n",
    "\n",
    "            if t1lad_obs_z > 0.00001:\n",
    "                pt1lad = len(enr_dat_t1lad.query('(TE_class == @te_class) and (n_intersections > @t1lad_ints)')) /\\\n",
    "                len(enr_dat_t1lad.query('(TE_class == @te_class)'))\n",
    "            else:\n",
    "                pt1lad = len(enr_dat_t1lad.query('(TE_class == @te_class) and (n_intersections < @t1lad_ints)')) /\\\n",
    "                len(enr_dat_t1lad.query('(TE_class == @te_class)'))\n",
    "\n",
    "            zscores.append(t1lad_obs_z)\n",
    "            pvalues.append(pt1lad)\n",
    "            lad_cts.append(ct)\n",
    "            enr_elements.append(te_class)\n",
    "            ladtypes.append('T1-LAD')\n",
    "\n",
    "            # calculate z-score for T2-LAD\n",
    "\n",
    "            t2lad_sd = np.std(enr_dat_t2lad.query('TE_class == @te_class')['n_intersections'])\n",
    "            t2lad_ints = actual_ints_t2lads[te_class]\n",
    "            t2lad_obs_z = (t2lad_ints - \\\n",
    "                         np.mean(enr_dat_t2lad.query('TE_class == @te_class')['n_intersections'])) / t2lad_sd\n",
    "\n",
    "            # calculate empirical p-value for T2-LAD\n",
    "\n",
    "            if t2lad_obs_z > 0.00001:\n",
    "                pt2lad = len(enr_dat_t2lad.query('(TE_class == @te_class) and (n_intersections > @t2lad_ints)')) /\\\n",
    "                len(enr_dat_t2lad.query('(TE_class == @te_class)'))\n",
    "            else:\n",
    "                pt2lad = len(enr_dat_t2lad.query('(TE_class == @te_class) and (n_intersections < @t2lad_ints)')) /\\\n",
    "                len(enr_dat_t2lad.query('(TE_class == @te_class)'))\n",
    "\n",
    "            zscores.append(t2lad_obs_z)\n",
    "            pvalues.append(pt2lad)\n",
    "            lad_cts.append(ct)\n",
    "            enr_elements.append(te_class)\n",
    "            ladtypes.append('T2-LAD')     \n",
    "\n",
    "    zpdf = pd.DataFrame({\n",
    "        'p_value':pvalues,\n",
    "        'zscore':zscores,\n",
    "        'LAD_type':ladtypes,\n",
    "        'class':enr_elements,\n",
    "        'celltype':lad_cts\n",
    "    }).dropna().query('zscore != \"-inf\"').query('zscore != \"inf\"')\n",
    "\n",
    "    zpdf['-log10p'] = -np.log10(zpdf['p_value'])\n",
    "    zpdf['-log10p'] = zpdf['-log10p'].replace(np.inf, 1.0)\n",
    "    return zpdf\n",
    "        \n",
    "def make_enrichment_df(feature_name, lad_or_kdd, cts):\n",
    "    # df for merged plot\n",
    "\n",
    "    zscores = []\n",
    "    pvalues = []\n",
    "    domain_cts = []\n",
    "    classes = []\n",
    "    domaintypes = []\n",
    "\n",
    "    for ct in cts:\n",
    "\n",
    "        t1dat = pd.read_csv(f'{lad_dat}/{feature_name}/{feature_name}_{ct}_T1{lad_or_kdd}s_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "        t2dat = pd.read_csv(f'{lad_dat}/{feature_name}/{feature_name}_{ct}_T2{lad_or_kdd}s_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "        \n",
    "        class_dats = {\n",
    "            'T1':t1dat,\n",
    "            'T2':t2dat\n",
    "        }\n",
    "\n",
    "        with open(f'{lad_dat}/{feature_name}/{feature_name}_{ct}_T1{lad_or_kdd}s_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t1 = json.load(f)\n",
    "\n",
    "        if actual_ints_t1 == 0:\n",
    "            actual_ints_t1 = 0.00001\n",
    "\n",
    "        with open(f'{lad_dat}/{feature_name}/{feature_name}_{ct}_T2{lad_or_kdd}s_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t2 = json.load(f)\n",
    "\n",
    "        if actual_ints_t2 == 0:\n",
    "            actual_ints_t2 = 0.00001\n",
    "\n",
    "        for dat_class in t1dat['feature_class'].unique():\n",
    "            \n",
    "            for class_type in ['T1', 'T2']:\n",
    "                \n",
    "                dat = class_dats[class_type]\n",
    "\n",
    "                # calculate z-score\n",
    "\n",
    "                sd = np.std(dat.query('feature_class == @dat_class')['n_intersections'])\n",
    "                if class_type == 'T1':\n",
    "                    ints = actual_ints_t1[dat_class]\n",
    "                else:\n",
    "                    ints = actual_ints_t2[dat_class]\n",
    "                obs_z = (ints - \\\n",
    "                         np.mean(dat.query('feature_class == @dat_class')['n_intersections'])) / sd\n",
    "\n",
    "                # calculate empirical p-value\n",
    "                \n",
    "                if obs_z > 0.00001:\n",
    "                    pval = len(dat.query('(feature_class == @dat_class) and (n_intersections > @ints)')) /\\\n",
    "                    len(t1dat.query('(feature_class == @dat_class)'))\n",
    "                else:\n",
    "                    pval = len(dat.query('(feature_class == @dat_class) and (n_intersections < @ints)')) /\\\n",
    "                    len(dat.query('(feature_class == @dat_class)'))\n",
    "\n",
    "                zscores.append(obs_z)\n",
    "                pvalues.append(pval)\n",
    "                domain_cts.append(ct)\n",
    "                classes.append(dat_class)\n",
    "                domaintypes.append(f'{class_type}-{lad_or_kdd}')\n",
    "\n",
    "    zpdf = pd.DataFrame({\n",
    "        'p_value':pvalues,\n",
    "        'zscore':zscores,\n",
    "        f'{lad_or_kdd}_type':domaintypes,\n",
    "        'class':classes,\n",
    "        'celltype':domain_cts\n",
    "    }).dropna().query('zscore != \"-inf\"').query('zscore != \"inf\"')\n",
    "\n",
    "    zpdf['p_value'] = zpdf['p_value'].replace(0.0, 0.0001)\n",
    "    zpdf['-log10p'] = -np.log10(zpdf['p_value'])\n",
    "    zpdf['-log10p'] = zpdf['-log10p'].replace(np.inf, 1.0)  \n",
    "    return(zpdf)\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "cts = ['Liver', 'CardiacMyocytes', 'EndoProgenitor', 'D5Midbrain',\n",
    "       'H9ESC', 'Epicardium', 'MidHindgut', 'ParaxMesoderm',\n",
    "       'EarlySomite', 'DefEctoderm', 'BorderEctoderm', 'D4Artery', 'APS']\n",
    "\n",
    "# enrichment df for genes\n",
    "zpdf_genes = get_enrichment_df(path_to_gene_results, 'genes')\n",
    "zpdf_genes['TE_class'] = 'Genes'\n",
    "\n",
    "# ATAC\n",
    "\n",
    "def load_atac(f):\n",
    "    return(BedTool.from_dataframe(pd.read_table(f'{path_to_atac}/GSE85330/{f}',\n",
    "                                                            header=None, names=['chrom','start','end','name',\n",
    "                                                                               'score','strand','score2',\n",
    "                                                                               'score3','score4','score5'])[['chrom','start','end']]).sort().liftover('/pollard/data/functional_genomics/liftOver/hg19ToHg38.over.chain.gz').sort())\n",
    "\n",
    "h9esc_atac_peaks_rep1 = load_atac('GSM2264826_H9_0_1.filterBL.bed.gz')\n",
    "h9esc_atac_peaks_rep2 = load_atac('GSM2264827_H9_0_2.filterBL.bed.gz')\n",
    "\n",
    "h9esc_atac_peaks = h9esc_atac_peaks_rep1.intersect(h9esc_atac_peaks_rep2).sort().merge()\n",
    "\n",
    "mes_atac_peaks_rep1 = load_atac('GSM2264828_H9_2_1.filterBL.bed.gz')\n",
    "mes_atac_peaks_rep2 = load_atac('GSM2264829_H9_2_2.filterBL.bed.gz')\n",
    "\n",
    "mes_atac_peaks = mes_atac_peaks_rep1.intersect(mes_atac_peaks_rep2).sort().merge()\n",
    "\n",
    "cm_atac_peaks_rep1 = load_atac('GSM2264832_H9_30_1.filterBL.bed.gz')\n",
    "cm_atac_peaks_rep2 = load_atac('GSM2264833_H9_30_2.filterBL.bed.gz')\n",
    "\n",
    "cm_atac_peaks = cm_atac_peaks_rep1.intersect(cm_atac_peaks_rep2).sort().merge()\n",
    "\n",
    "atac_peaks = {}\n",
    "\n",
    "atac_peaks['H9ESC'] = h9esc_atac_peaks\n",
    "atac_peaks['ParaxMesoderm'] = mes_atac_peaks\n",
    "atac_peaks['CardiacMyocytes'] = cm_atac_peaks\n",
    "\n",
    "atac_kdd_zpdf = make_enrichment_df('ATAC','KDD',atac_peaks.keys())\n",
    "atac_lad_zpdf = make_enrichment_df('ATAC','LAD',atac_peaks.keys())\n",
    "\n",
    "bcomp_cts = ['H9ESC','CardiacMyocytes']\n",
    "\n",
    "bcomp_kdd_zpdf = make_enrichment_df('bcomp', 'KDD', bcomp_cts)\n",
    "bcomp_lad_zpdf = make_enrichment_df('bcomp', 'LAD', bcomp_cts)\n",
    "\n",
    "cts = ['Liver', 'CardiacMyocytes', 'EndoProgenitor', 'D5Midbrain',\n",
    "       'H9ESC', 'Epicardium', 'MidHindgut', 'ParaxMesoderm',\n",
    "       'EarlySomite', 'DefEctoderm', 'BorderEctoderm', 'D4Artery', 'APS']\n",
    "\n",
    "zpdf_genes['class'] = 'Genes'\n",
    "\n",
    "zpdf = pd.concat([zpdf_genes, atac_lad_zpdf, bcomp_lad_zpdf], sort=False)\n",
    "\n",
    "ct_replace = {\n",
    "    'CardiacMyocytes':'Cardiac\\nmyocytes',\n",
    "    'EarlySomite':'Early somite',\n",
    "    'H9ESC':'ESCs',\n",
    "    'ParaxMesoderm':'Paraxial\\nmesoderm',\n",
    "    'DefEctoderm':'Definitive\\nectoderm',\n",
    "    'MidHindgut':'Mid-hindgut',\n",
    "    'D5Midbrain':'Midbrain',\n",
    "    'Liver':'Liver progenitors',\n",
    "    'BorderEctoderm':'Border ectoderm',\n",
    "    'EndoProgenitor':'Endothelial\\nprogenitors',\n",
    "    'D4Artery':'Artery progenitors',\n",
    "    'APS':'Anterior primitive\\nstreak'\n",
    "}\n",
    "\n",
    "zpdf_genes['celltype'].unique()\n",
    "\n",
    "zpdf['celltype'].unique()\n",
    "\n",
    "f, ax = plt.subplots(figsize=(4,3))\n",
    "sns.scatterplot(y='class', x='zscore', size='-log10p', data=zpdf.replace(ct_replace), hue='celltype',\n",
    "               style='LAD_type', palette='colorblind', ax=ax, sizes=(20, 200), alpha=0.6, linewidth=0)\n",
    "plt.legend(ncol=2, bbox_to_anchor=(1, 1), frameon=False)\n",
    "plt.axvline(0, linestyle='--', color='grey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e5e1b96",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LAD_analysis",
   "language": "python",
   "name": "lad_analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
