{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29e697c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0a2a3a0d",
   "metadata": {},
   "source": [
    "# Feature enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "836f6f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_tes = 'dat/TEs'\n",
    "path_to_rt_results = 'dat/repdomains'\n",
    "path_to_sc_dat = 'kind_et_al/2015'\n",
    "hg19_to_hg38 = 'liftOver/hg19ToHg38.over.chain.gz'\n",
    "path_to_lad_calls = 'LADs_KDDs/hmm_calls_3states'\n",
    "chromhmm_path = 'chromHMM/hg38'\n",
    "\n",
    "# load transposable elements\n",
    "\n",
    "cts = ['Liver', 'CardiacMyocytes', 'EndoProgenitor', 'D5Midbrain',\n",
    "       'H9ESC', 'Epicardium', 'MidHindgut', 'ParaxMesoderm',\n",
    "       'EarlySomite', 'DefEctoderm', 'BorderEctoderm', 'D4Artery', 'APS']\n",
    "\n",
    "def get_enrichment_df(enr_results_path, enr_feature):\n",
    "    \"\"\"\n",
    "    Collects enrichment results (pre-computed) into a dataframe for plotting\n",
    "    \"\"\"\n",
    "    zscores = []\n",
    "    pvalues = []\n",
    "    lad_cts = []\n",
    "    enr_elements = []\n",
    "    ladtypes = []\n",
    "\n",
    "    for ct in cts:\n",
    "#         print(ct)\n",
    "\n",
    "        enr_dat_t1lad = pd.read_csv(f'{enr_results_path}/{enr_feature}_{ct}_T1LADs_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "        enr_dat_t2lad = pd.read_csv(f'{enr_results_path}/{enr_feature}_{ct}_T2LADs_enrichment_results.tsv',\n",
    "                            sep='\\t').replace(0.0, 0.00001)\n",
    "\n",
    "        with open(f'{enr_results_path}/{enr_feature}_{ct}_T1LADs_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t1lads = json.load(f)\n",
    "\n",
    "        if actual_ints_t1lads == 0:\n",
    "            actual_ints_t1lads = 0.00001\n",
    "\n",
    "        with open(f'{enr_results_path}/{enr_feature}_{ct}_T2LADs_actual_intersections.json', 'r') as f:\n",
    "            actual_ints_t2lads = json.load(f)\n",
    "\n",
    "        if actual_ints_t2lads == 0:\n",
    "            actual_ints_t2lads = 0.00001\n",
    "\n",
    "        # this loop was originally developed for TEs, but now \n",
    "        # works with all the features we assessed for enrichment\n",
    "        for te_class in enr_dat_t1lad['TE_class'].unique():\n",
    "\n",
    "            # calculate z-score for T1-LAD\n",
    "\n",
    "            t1lad_sd = np.std(enr_dat_t1lad.query('TE_class == @te_class')['n_intersections'])\n",
    "            t1lad_ints = actual_ints_t1lads[te_class]\n",
    "            t1lad_obs_z = (t1lad_ints - \\\n",
    "                         np.mean(enr_dat_t1lad.query('TE_class == @te_class')['n_intersections'])) / t1lad_sd\n",
    "\n",
    "            # calculate empirical p-value for T1-LAD\n",
    "\n",
    "            if t1lad_obs_z > 0.00001:\n",
    "                pt1lad = len(enr_dat_t1lad.query('(TE_class == @te_class) and (n_intersections > @t1lad_ints)')) /\\\n",
    "                len(enr_dat_t1lad.query('(TE_class == @te_class)'))\n",
    "            else:\n",
    "                pt1lad = len(enr_dat_t1lad.query('(TE_class == @te_class) and (n_intersections < @t1lad_ints)')) /\\\n",
    "                len(enr_dat_t1lad.query('(TE_class == @te_class)'))\n",
    "\n",
    "            zscores.append(t1lad_obs_z)\n",
    "            pvalues.append(pt1lad)\n",
    "            lad_cts.append(ct)\n",
    "            enr_elements.append(te_class)\n",
    "            ladtypes.append('T1-LAD')\n",
    "\n",
    "            # calculate z-score for T2-LAD\n",
    "\n",
    "            t2lad_sd = np.std(enr_dat_t2lad.query('TE_class == @te_class')['n_intersections'])\n",
    "            t2lad_ints = actual_ints_t2lads[te_class]\n",
    "            t2lad_obs_z = (t2lad_ints - \\\n",
    "                         np.mean(enr_dat_t2lad.query('TE_class == @te_class')['n_intersections'])) / t2lad_sd\n",
    "\n",
    "            # calculate empirical p-value for T2-LAD\n",
    "\n",
    "            if t2lad_obs_z > 0.00001:\n",
    "                pt2lad = len(enr_dat_t2lad.query('(TE_class == @te_class) and (n_intersections > @t2lad_ints)')) /\\\n",
    "                len(enr_dat_t2lad.query('(TE_class == @te_class)'))\n",
    "            else:\n",
    "                pt2lad = len(enr_dat_t2lad.query('(TE_class == @te_class) and (n_intersections < @t2lad_ints)')) /\\\n",
    "                len(enr_dat_t2lad.query('(TE_class == @te_class)'))\n",
    "\n",
    "            zscores.append(t2lad_obs_z)\n",
    "            pvalues.append(pt2lad)\n",
    "            lad_cts.append(ct)\n",
    "            enr_elements.append(te_class)\n",
    "            ladtypes.append('T2-LAD')     \n",
    "\n",
    "    zpdf = pd.DataFrame({\n",
    "        'p_value':pvalues,\n",
    "        'zscore':zscores,\n",
    "        'LAD_type':ladtypes,\n",
    "        'TE_class':enr_elements,\n",
    "        'celltype':lad_cts\n",
    "    }).dropna().query('zscore != \"-inf\"').query('zscore != \"inf\"')\n",
    "\n",
    "    zpdf['-log10p'] = -np.log10(zpdf['p_value'])\n",
    "    zpdf['-log10p'] = zpdf['-log10p'].replace(np.inf, 1.0)\n",
    "    return zpdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c602a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment df for TEs\n",
    "zpdf_tes = get_enrichment_df(path_to_tes, 'TEs_by_fam')\n",
    "\n",
    "# enrichment df for replication timing domains\n",
    "zpdf_rt = get_enrichment_df(path_to_rt_results, 'RTs')\n",
    "\n",
    "# find TE families with same enrichment pattern in all celltypes\n",
    "# e.g. T1 and T2-LADs all enriched, or T1-LADs enriched, T2-LADs depleted, or vice versa\n",
    "\n",
    "for_include = []\n",
    "        \n",
    "te_cs = zpdf_tes['TE_class'].unique()\n",
    "for te in zpdf_tes.query('TE_class in @te_cs')['TE_class'].unique():\n",
    "    if all(zpdf_tes.query('TE_class == @te')['zscore'] > 0) or all(zpdf_tes.query('TE_class == @te')['zscore'] < 0):\n",
    "        for_include.append(te)\n",
    "    elif all(zpdf_tes.query('TE_class == @te and LAD_type == \"T1-LAD\"')['zscore'] > 0) and all(zpdf_tes.query('TE_class == @te and LAD_type == \"T2-LAD\"')['zscore'] < 0):\n",
    "        for_include.append(te)\n",
    "    elif all(zpdf_tes.query('TE_class == @te and LAD_type == \"T1-LAD\"')['zscore'] < 0) and all(zpdf_tes.query('TE_class == @te and LAD_type == \"T2-LAD\"')['zscore'] > 0):\n",
    "        for_include.append(te)\n",
    "  \n",
    "exclude = ['Unknown']\n",
    "for_include = [ x for x in for_include if not (x.endswith('?') or x in exclude) ]\n",
    "\n",
    "zpdf_tog = pd.concat([zpdf_tes.query('TE_class in @for_include').copy(), zpdf_rt], sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb1b5339",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 4))\n",
    "g = sns.scatterplot(x='zscore', y='TE_class', data=zpdf_tog, size='-log10p', \n",
    "                   hue='celltype', style='LAD_type', legend='brief',\n",
    "                   palette='colorblind', sizes=(20, 200))\n",
    "plt.xticks(rotation=90)\n",
    "plt.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\")\n",
    "plt.grid(b=None)\n",
    "plt.ylabel('')\n",
    "plt.xlabel('Z-score')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70fd694a",
   "metadata": {},
   "source": [
    "# CTCF at boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efe17b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add CTCF\n",
    "\n",
    "def load_ctcf_dat(path, rtype):\n",
    "    ctcf = pd.read_table(path,\n",
    "                         skiprows=1, header=0, low_memory=False).replace(np.nan, 0.0).replace('--',0.0)\n",
    "    ctcf.columns = ['CTCF',rtype] + list(range(ctcf.shape[1]))[:-2]\n",
    "    ctcf.columns = ['CTCF',rtype] + list(range(ctcf.shape[1]))[:-2]\n",
    "\n",
    "    dfs = []\n",
    "\n",
    "    for ix, row in ctcf.iterrows():\n",
    "        CTCF_dset = row['CTCF']\n",
    "        r_dset = row[rtype]\n",
    "\n",
    "        dat = pd.DataFrame(row[2:].T)\n",
    "        dat.columns=['CTCF_sig']\n",
    "        dat['CTCF'] = CTCF_dset.split('_')[0]\n",
    "        dat[rtype] = r_dset.split('_')[0]\n",
    "        dat['bin'] = dat.index\n",
    "\n",
    "        dfs.append(dat)\n",
    "\n",
    "\n",
    "    ctcf_tog = pd.concat(dfs, sort=False)\n",
    "    ctcf_tog['CTCF_sig'] = ctcf_tog['CTCF_sig'].astype(float)\n",
    "    return(ctcf_tog)\n",
    "\n",
    "max_bin = 200\n",
    "\n",
    "# T1 to non\n",
    "\n",
    "non_to_t1_all = []\n",
    "\n",
    "for ct in ['escs','parax_mesoderm','cms']:\n",
    "    t1_to_non = load_ctcf_dat(f'ctcf_reanalysis/t1_to_non_{ct}_for_plot.tsv','T1_to_non').query('bin < @max_bin')\n",
    "    rev_bins = t1_to_non['bin'].tolist()\n",
    "    rev_bins.reverse()\n",
    "    t1_to_non['bin'] = rev_bins\n",
    "    non_to_t1 = load_ctcf_dat(f'ctcf_reanalysis/non_to_t1_{ct}_for_plot.tsv','T1_to_non').query('bin < @max_bin')\n",
    "    t1_to_non_tog = pd.concat([t1_to_non, non_to_t1], sort=False).sort_values(by='bin', ascending=False)\n",
    "    t1_to_non_tog['cat'] = 'T1_to_non'\n",
    "    t1_to_non_tog['ct'] = ct\n",
    "    non_to_t1_all.append(t1_to_non_tog)\n",
    "    \n",
    "non_to_t1_all_cts = pd.concat(non_to_t1_all, sort=False)\n",
    "\n",
    "# T2 to non\n",
    "\n",
    "non_to_t2_all = []\n",
    "\n",
    "for ct in ['escs','parax_mesoderm','cms']:\n",
    "    t2_to_non = load_ctcf_dat(f'ctcf_reanalysis/t2_to_non_{ct}_for_plot.tsv','T2_to_non').query('bin < @max_bin')\n",
    "    rev_bins = t2_to_non['bin'].tolist()\n",
    "    rev_bins.reverse()\n",
    "    t2_to_non['bin'] = rev_bins\n",
    "    non_to_t2 = load_ctcf_dat(f'ctcf_reanalysis/non_to_t2_{ct}_for_plot.tsv','T2_to_non').query('bin < @max_bin')\n",
    "    t2_to_non_tog = pd.concat([t2_to_non, non_to_t2], sort=False).sort_values(by='bin', ascending=False)\n",
    "    t2_to_non_tog['cat'] = 'T2_to_non'\n",
    "    t2_to_non_tog['ct'] = ct\n",
    "    non_to_t2_all.append(t2_to_non_tog)\n",
    "    \n",
    "non_to_t2_all_cts = pd.concat(non_to_t2_all, sort=False)\n",
    "\n",
    "# T2 to T1\n",
    "\n",
    "t2_to_t1_all = []\n",
    "\n",
    "for ct in ['escs','parax_mesoderm','cms']:\n",
    "    t1_to_t2 = load_ctcf_dat(f'ctcf_reanalysis/t1_to_t2_{ct}_for_plot.tsv','T2_to_T1').query('bin < @max_bin')\n",
    "    rev_bins = t1_to_t2['bin'].tolist()\n",
    "    rev_bins.reverse()\n",
    "    t1_to_t2['bin'] = rev_bins\n",
    "    t2_to_t1 = load_ctcf_dat(f'ctcf_reanalysis/t2_to_t1_{ct}_for_plot.tsv','T2_to_T1').query('bin < @max_bin')\n",
    "    t2_to_t1_tog = pd.concat([t2_to_t1, t1_to_t2], sort=False).sort_values(by='bin', ascending=False)\n",
    "    t2_to_t1_tog['cat'] = 'T2_to_t1'\n",
    "    t2_to_t1_tog['ct'] = ct\n",
    "    t2_to_t1_all.append(t2_to_t1_tog)\n",
    "    \n",
    "t2_to_t1_all_cts = pd.concat(t2_to_t1_all, sort=False)\n",
    "\n",
    "ctr = {\n",
    "    'Cardiac myocytes':'cms',\n",
    "    'Paraxial mesoderm':'parax_mesoderm',\n",
    "    'ESCs':'escs'\n",
    "}\n",
    "\n",
    "# characterize number of types of regions per cell type\n",
    "\n",
    "cats = []\n",
    "cts = []\n",
    "n_bs = []\n",
    "\n",
    "for ct in ['ESCs','Paraxial mesoderm','Cardiac myocytes']:\n",
    "    ctr_ = ctr[ct]\n",
    "    t1_to_non = pd.concat([BedTool(f'ctcf_reanalysis/t1_to_non_{ctr_}.bed').to_dataframe(),\n",
    "                              BedTool(f'ctcf_reanalysis/non_to_t1_{ctr_}.bed').to_dataframe()],\n",
    "                            sort=False)\n",
    "    cats.append('nonLAD to T1-LAD')\n",
    "    cts.append(ct)\n",
    "    n_bs.append(len(t1_to_non))\n",
    "    \n",
    "    t2_to_non = pd.concat([BedTool(f'ctcf_reanalysis/t2_to_non_{ctr_}.bed').to_dataframe(),\n",
    "                              BedTool(f'ctcf_reanalysis/non_to_t2_{ctr_}.bed').to_dataframe()],\n",
    "                            sort=False)\n",
    "    cats.append('nonLAD to T2-LAD')\n",
    "    cts.append(ct)\n",
    "    n_bs.append(len(t2_to_non))\n",
    "    \n",
    "    t1_to_t2 = pd.concat([BedTool(f'ctcf_reanalysis/t2_to_t1_{ctr_}.bed').to_dataframe(),\n",
    "                              BedTool(f'ctcf_reanalysis/t1_to_t2_{ctr_}.bed').to_dataframe()],\n",
    "                            sort=False)\n",
    "    cats.append('T2-LAD to T1-LAD')\n",
    "    cts.append(ct)\n",
    "    n_bs.append(len(t1_to_t2))\n",
    "    \n",
    "n_doms_df = pd.DataFrame({\n",
    "    'category':cats,\n",
    "    'cell_type':cts,\n",
    "    'n_boundaries':n_bs\n",
    "})\n",
    "\n",
    "\n",
    "# CTCF nonLAD --> T2-LAD\n",
    "\n",
    "axC = plt.subplot(gs[1, 1])\n",
    "c = sns.lineplot(x='bin', y='CTCF_sig', data=non_to_t2_all_cts.replace({'escs':'ESCs',\n",
    "                                                                       'cms':'Cardiac myocytes',\n",
    "                                                                       'parax_mesoderm':'Paraxial mesoderm'}), \n",
    "                 hue='ct', ci='sd', alpha=0.5, estimator='mean', \n",
    "                 hue_order=['ESCs','Paraxial mesoderm','Cardiac myocytes'],\n",
    "                ax=axC)\n",
    "\n",
    "\n",
    "\n",
    "c.set(ylabel='CTCF binding [rpkm]', xlabel='genomic position')\n",
    "axC.legend(frameon=False, bbox_to_anchor = (1., -0.2))\n",
    "axC.set_xticks([0, 100, 200])\n",
    "axC.set_xticklabels(['-50kb','boundary','+50kb'])\n",
    "axC.axvline(100, linestyle='--', color='grey')\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "axC.annotate('',\n",
    "            xy=(0.1, .98), xycoords='axes fraction',\n",
    "            xytext=(0.4, .98), textcoords='axes fraction',\n",
    "            arrowprops=dict(facecolor='grey', shrink=0.05, alpha=0.5),\n",
    "            horizontalalignment='right', verticalalignment='top')\n",
    "axC.text(0.3, 1.1, 'nonLAD', fontsize=12, transform=axC.transAxes,\n",
    "         horizontalalignment='center',\n",
    "         verticalalignment='center')\n",
    "axC.annotate('',\n",
    "            xy=(0.9, .98), xycoords='axes fraction',\n",
    "            xytext=(0.6, .98), textcoords='axes fraction',\n",
    "            arrowprops=dict(facecolor='grey', shrink=0.05, alpha=0.5),\n",
    "            horizontalalignment='right', verticalalignment='top')\n",
    "axC.text(0.7, 1.1, 'T2-LAD', fontsize=12, transform=axC.transAxes,\n",
    "         horizontalalignment='center',\n",
    "         verticalalignment='center')\n",
    "\n",
    "# CTCF T2-LAD --> T1-LAD\n",
    "\n",
    "axD = plt.subplot(gs[1, 2])\n",
    "d = sns.lineplot(x='bin', y='CTCF_sig', data=t2_to_t1_all_cts.replace({'escs':'ESCs',\n",
    "                                                                       'cms':'Cardiac myocytes',\n",
    "                                                                       'parax_mesoderm':'Paraxial mesoderm'}), \n",
    "                 hue='ct', ci='sd', alpha=0.5, estimator='mean', \n",
    "                 hue_order=['ESCs','Paraxial mesoderm','Cardiac myocytes'],\n",
    "                ax=axD)\n",
    "\n",
    "\n",
    "\n",
    "d.set(ylabel='CTCF binding [rpkm]', xlabel='genomic position')\n",
    "axD.legend(frameon=False, bbox_to_anchor = (1., -.2))\n",
    "axD.set_xticks([0, 100, 200])\n",
    "axD.set_xticklabels(['-50kb','boundary','+50kb'])\n",
    "axD.axvline(100, linestyle='--', color='grey')\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "axD.annotate('',\n",
    "            xy=(0.1, .98), xycoords='axes fraction',\n",
    "            xytext=(0.4, .98), textcoords='axes fraction',\n",
    "            arrowprops=dict(facecolor='grey', shrink=0.05, alpha=0.5),\n",
    "            horizontalalignment='right', verticalalignment='top')\n",
    "axD.text(0.3, 1.1, 'T2-LAD', fontsize=12, transform=axD.transAxes,\n",
    "         horizontalalignment='center',\n",
    "         verticalalignment='center')\n",
    "axD.annotate('',\n",
    "            xy=(0.9, .98), xycoords='axes fraction',\n",
    "            xytext=(0.6, .98), textcoords='axes fraction',\n",
    "            arrowprops=dict(facecolor='grey', shrink=0.05, alpha=0.5),\n",
    "            horizontalalignment='right', verticalalignment='top')\n",
    "axD.text(0.7, 1.1, 'T1-LAD', fontsize=12, transform=axD.transAxes,\n",
    "         horizontalalignment='center',\n",
    "         verticalalignment='center')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd7d415c",
   "metadata": {},
   "source": [
    "# Single cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51dd5de7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load Kind et al. 2015 single cell data\n",
    "dat = pd.read_csv(f'{path_to_sc_dat}/GSE68260_Clone.5-5.1N.OE_LP150415.txt',\n",
    "                 sep='\\t')\n",
    "\n",
    "# determine percentage of cells with LB1 association at each locus\n",
    "dat['perc_cells_LB1'] = 100.0 * ((dat[dat.columns[4:-1]] > 0.0).sum(1) / 172.0 )\n",
    "\n",
    "# liftOver to hg38 from hg19\n",
    "bed_hg38 = BedTool.from_dataframe(dat[['seqnames','start','end','perc_cells_LB1']]).liftover(hg19_to_hg38).sort()\n",
    "\n",
    "def assign_categories(dat, cat_names):\n",
    "    \"\"\"\n",
    "    Assign categories to specific HMM states\n",
    "    based on the mean ChIP-seq coverage\n",
    "    in bins assigned to that state.\n",
    "    \n",
    "    dat is the binned data with HMM predicted states\n",
    "    generated by lad_utils.hmm_predict()\n",
    "    \n",
    "    cat_names should be provided based on which should be\n",
    "    assigned to sequecing coverage level in ascending order.\n",
    "    \"\"\"\n",
    "    n_states = len(dat['hmm_pred'].unique())\n",
    "    n_cats = len(cat_names)\n",
    "    if n_states != n_cats:\n",
    "        print(f'{n_states} in data, but {n_cats} provided. '\\\n",
    "              f'Number of categories must match the number '\\\n",
    "              f'of HMM states.')\n",
    "    else:\n",
    "        \n",
    "        cols = ['score0','score1'] # each cell type has 2 replicates\n",
    "        # cols = ['score']\n",
    "        \n",
    "        states = []\n",
    "        means = []\n",
    "        \n",
    "        for state in dat['hmm_pred'].unique():\n",
    "            states.append(state)\n",
    "            means.append(np.mean(dat.query(f'hmm_pred == {state}')[cols].mean()))\n",
    "           \n",
    "        # sort means and associated states, ascending\n",
    "        sorted_means = sorted(means)\n",
    "        means_to_states = dict(zip(means, states))\n",
    "        \n",
    "        sorted_states = []\n",
    "        \n",
    "        for val in sorted_means:\n",
    "            sorted_states.append(means_to_states[val])\n",
    "    \n",
    "        names_to_states = dict(zip(sorted_states, cat_names))\n",
    "        dat['category'] = dat['hmm_pred'].replace(names_to_states)\n",
    "        return dat\n",
    "\n",
    "n_states = 3\n",
    "\n",
    "# Paraxial mesoderm is the closest cell type we have\n",
    "lb1_dat = pd.read_table(f'{path_to_lad_calls}/ParaxMesoderm.tsv',\n",
    "                       header=0, names=['chrom','start','stop','score0','score1','hmm_pred'])\n",
    "lb1_dat = assign_categories(lb1_dat.copy(), cat_names=['nonLAD','T2-LAD','T1-LAD'])\n",
    "lb1_dat['id'] = lb1_dat.index\n",
    "lb1_dat_bed = BedTool.from_dataframe(lb1_dat[['chrom','start','stop','id']]).sort()\n",
    "\n",
    "# intersect LB1 bins with single cell data\n",
    "lb1_scdat = lb1_dat_bed.intersect(bed_hg38, wa=True, wb=True).to_dataframe()[['name','thickEnd']]\n",
    "lb1_scdat.columns = ['id','perc_cells_LB1']\n",
    "\n",
    "# merge back to LB1 data\n",
    "lb1_w_sc = lb1_dat.merge(lb1_scdat, on='id', how='left').fillna(0.0)\n",
    "\n",
    "sns.violinplot(y='category', x='perc_cells_LB1', \n",
    "               data=lb1_w_sc, #sc_dat from LADs and LADlites vs. sc data\n",
    "              order=['nonLAD','T2-LAD','T1-LAD'],\n",
    "              scale='count', linewidth=1, inner='box')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc096218",
   "metadata": {},
   "source": [
    "# chromHMM results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81604fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cts = ['H9ESC','CardiacMyocytes', 'Liver', 'ParaxMesoderm']\n",
    "\n",
    "esc_chromhmm = pd.read_table(f'{chromhmm_path}/E008_25_imputed12marks_hg38lift_dense.bed', \n",
    "                            skiprows=1, header=None, usecols=range(4), names=['chrom','start','stop','state'])\n",
    "\n",
    "cm_chromhmm = pd.read_table(f'{chromhmm_path}/E095_25_imputed12marks_hg38lift_dense.bed', \n",
    "                            skiprows=1, header=None, usecols=range(4), names=['chrom','start','stop','state'])\n",
    "\n",
    "liver_chromhmm = pd.read_table(f'{chromhmm_path}/E066_25_imputed12marks_hg38lift_dense.bed', \n",
    "                            skiprows=1, header=None, usecols=range(4), names=['chrom','start','stop','state'])\n",
    "\n",
    "mesoderm_chromhmm = pd.read_table(f'{chromhmm_path}/E013_25_imputed12marks_hg38lift_dense.bed', \n",
    "                            skiprows=1, header=None, usecols=range(4), names=['chrom','start','stop','state'])\n",
    "\n",
    "chromhmms = {}\n",
    "\n",
    "chromhmms['H9ESC'] = esc_chromhmm\n",
    "chromhmms['CardiacMyocytes'] = cm_chromhmm\n",
    "chromhmms['Liver'] = liver_chromhmm\n",
    "chromhmms['ParaxMesoderm'] = mesoderm_chromhmm\n",
    "\n",
    "ct_lb1 = {}\n",
    "\n",
    "for ct in cts:\n",
    "    lb1_dat = pd.read_table(f'{path_to_lad_calls}/{ct}.tsv',\n",
    "                           header=0, names=['chrom','start','stop','score0','score1','hmm_pred'])\n",
    "    lb1_dat = assign_categories(lb1_dat.copy(), cat_names=['nonLAD','T2-LAD','T1-LAD'])\n",
    "    lb1_dat['id'] = lb1_dat.index\n",
    "    \n",
    "    ct_lb1[ct] = lb1_dat\n",
    "    \n",
    "    \n",
    "cts_for_plot = []\n",
    "cats_for_plot = []\n",
    "odds_ratios = []\n",
    "pvals = []\n",
    "chromhmm_cat = []\n",
    "\n",
    "for ct in cts:\n",
    "    for lad_cat in ['T1-LAD','T2-LAD']:\n",
    "        lad_bed = BedTool.from_dataframe(ct_lb1[ct].query('category == @lad_cat')[['chrom','start','stop','id']].copy()).sort().set_chromsizes('hg38')\n",
    "\n",
    "        all_cats = {}\n",
    "\n",
    "        depleted = []\n",
    "        enriched = []\n",
    "\n",
    "        chromhmm = chromhmms[ct]\n",
    "        for cat in chromhmm['state'].unique():\n",
    "            boi = BedTool.from_dataframe(chromhmm.query('state == @cat').copy()).sort().set_chromsizes('hg38')\n",
    "            out = str(boi.fisher(lad_bed))\n",
    "            if float(out.split('\\n')[-2].split('\\t')[0]) < 0.01:\n",
    "                depleted.append(cat)\n",
    "            elif float(out.split('\\n')[-2].split('\\t')[1]) < 0.01:\n",
    "                enriched.append(cat)\n",
    "            pvals.append(float(out.split('\\n')[-2].split('\\t')[1]))\n",
    "            cts_for_plot.append(ct)\n",
    "            cats_for_plot.append(lad_cat)\n",
    "            odds_ratios.append(float(out.split('\\n')[-2].split('\\t')[3]))\n",
    "            chromhmm_cat.append(cat)\n",
    "            \n",
    "chromhmm_replace = {\n",
    "    '1_TssA':'Active TSS',\n",
    "    '2_PromU':'Promoter Upstream TSS',\n",
    "    '3_PromD1':'Promoter Downstream TSS 1',\n",
    "    '4_PromD2':'Promoter Downstream TSS 2',\n",
    "    '5_Tx5':\"Transcribed - 5' preferential\",\n",
    "    '6_Tx':'Strong transcription',\n",
    "    '7_Tx3':\"Transcribed - 3' preferential\",\n",
    "    '8_TxWk':'Weak transcription',\n",
    "    '9_TxReg':'Transcribed & regulatory (Prom/Enh)',\n",
    "    '10_TxEnh5':\"Transcribed 5' preferential and Enh\",\n",
    "    '11_TxEnh3':\"Transcribed 3' preferential and Enh\",\n",
    "    '12_TxEnhW':'Transcribed and Weak Enhancer',\n",
    "    '13_EnhA1':'Active Enhancer 1',\n",
    "    '14_EnhA2':'Active Enhancer 2',\n",
    "    '15_EnhAF':'Active Enhancer Flank',\n",
    "    '16_EnhW1':'Weak Enhancer 1',\n",
    "    '17_EnhW2':'Weak Enhancer 2',\n",
    "    '18_EnhAc':'Primary H3k27ac possible Enhancer',\n",
    "    '19_DNase':'Primary DNase',\n",
    "    '20_ZNF/Rpts':'ZNF genes & repeats',\n",
    "    '21_Het':'Heterochromatin',\n",
    "    '22_PromP':'Poised Promoter',\n",
    "    '23_PromBiv':'Bivalent Promoter',\n",
    "    '24_ReprPC':'Repressed Polycomb',\n",
    "    '25_Quies':'Quiescent/Low',\n",
    "    \"10_TxEnh5'\":\"Transcribed 5' preferential and Enh\",\n",
    "    \"11_TxEnh3'\":\"Transcribed 3' preferential and Enh\",\n",
    "    \"5_Tx5'\":\"Transcribed - 5' preferential\",\n",
    "    \"7_Tx3'\":\"Transcribed - 3' preferential\"\n",
    "}\n",
    "\n",
    "chromhmm_df_plot = pd.DataFrame({\n",
    "    'celltype':cts_for_plot,\n",
    "    'LAD_cat':cats_for_plot,\n",
    "    'odds_ratio':odds_ratios,\n",
    "    'pval':pvals,\n",
    "    'chromHMM':chromhmm_cat\n",
    "}).replace(np.nan, 1.0)\n",
    "\n",
    "chromhmm_df_plot = chromhmm_df_plot.replace(chromhmm_replace)\n",
    "chromhmm_df_plot['celltype & LAD category'] = chromhmm_df_plot['celltype'] + '_' + chromhmm_df_plot['LAD_cat']\n",
    "\n",
    "hm_chmm = chromhmm_df_plot[['celltype & LAD category','chromHMM','odds_ratio']].copy().pivot('celltype & LAD category','chromHMM','odds_ratio')\n",
    "\n",
    "\n",
    "chromhmm_df_plot['pval'] = chromhmm_df_plot['pval'].astype(float)\n",
    "chromhmm_df_plot['pval_sig'] = chromhmm_df_plot['pval'] < 0.01\n",
    "chromhmm_df_plot['pval_formatted'] = chromhmm_df_plot['pval_sig'].replace({True:'*', False:''})\n",
    "\n",
    "hm_chmm2 = chromhmm_df_plot[['celltype & LAD category','chromHMM','pval_formatted']].copy().pivot('celltype & LAD category','chromHMM','pval_formatted')\n",
    "\n",
    "from matplotlib import cm\n",
    "\n",
    "cb_pal4 = np.array(sns.color_palette(palette='colorblind')[:4])\n",
    "\n",
    "cb_pal = np.array(['indigo','mediumpurple'])\n",
    "\n",
    "lut = dict(zip(hm_chmm.index.str.split('_').str[1].unique(), cb_pal))\n",
    "row_colors = hm_chmm.index.str.split('_').str[1].map(lut)\n",
    "lut = {\n",
    "    'CardiacMyocytes':'tab:green',\n",
    "    'H9ESC':'tab:blue',\n",
    "    'Liver':'tab:cyan',\n",
    "    'ParaxMesoderm':'tab:orange'\n",
    "}\n",
    "row_colors2 = hm_chmm.index.str.split('_').str[0].map(lut)\n",
    "\n",
    "cmap = sns.clustermap(hm_chmm, \n",
    "               cmap='viridis', col_cluster=False, \n",
    "               row_cluster=True, \n",
    "               row_colors=[row_colors2, row_colors],\n",
    "              vmin=0.0, metric='cosine', figsize=(10,6),\n",
    "                     dendrogram_ratio = 0.1, cbar_pos=(0.01, .05, .03, .2),\n",
    "                     cbar_kws = {'label':'odds ratio'}, colors_ratio = 0.03,\n",
    "                      annot_kws={'rotation': 90},\n",
    "                     **{'annot':hm_chmm2, 'center':3}, fmt='s')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fa88de4",
   "metadata": {},
   "source": [
    "# H3K9me2 in LADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa33633",
   "metadata": {},
   "outputs": [],
   "source": [
    "density_dat = pd.read_csv('tables_out/all_dat_tog.tsv', sep='\\t').replace('EarlySomite','Early somite')\n",
    "\n",
    "density_dat['score_lb1'] = density_dat[['score0_lb1','score1_lb1']].mean(axis=1)\n",
    "density_dat['score_k9'] = density_dat[['score0_k9','score1_k9']].mean(axis=1)\n",
    "\n",
    "f, axB = plt.subplots(figsize=(6,8))\n",
    "\n",
    "b = sns.violinplot(x='score_h3k9me2', y='cell_type', data=density_dat, kind='violin',\n",
    "           hue='category_lb1',linewidth=0.5, inner=None, ax=axB,\n",
    "                   hue_order=['T1-LAD','T2-LAD','nonLAD'],\n",
    "          order=['ESCs','Mid-hindgut','Liver','Anterior primitive streak',\n",
    "                'Paraxial mesoderm','Early somite','Artery progenitors',\n",
    "                'Endothelial progenitors','Cardiac myocytes','Epicardium',\n",
    "                'Definitive ectoderm','Border ectoderm','Midbrain progenitors'],\n",
    "                  palette = mypal)\n",
    "plt.legend(frameon=False, title='KDD category', bbox_to_anchor=(1, 1.15),\n",
    "          ncol=1)\n",
    "# plt.xticks(rotation=45, ha='right')\n",
    "plt.gca().spines['top'].set_visible(False)\n",
    "plt.gca().spines['right'].set_visible(False)\n",
    "axB.set(ylabel='cell type', xlabel='H3K9me2 binding [log2(read counts)]')\n",
    "plt.tight_layout()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LAD_analysis",
   "language": "python",
   "name": "lad_analysis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
